### ISSUE TITLE
`no-unused-vars` error when using `eslint-plugin-mdx` without `eslint-plugin-react`.

### ISSUE THREAD

**Issue by chenxsan** [mdx-js/eslint-mdx#444]
### Initial checklist

- [X] I read the [support docs](https://mdxjs.com/community/support/)
- [X] I read the [contributing guide](https://mdxjs.com/community/contribute/)
- [X] I agree to follow the [code of conduct](https://github.com/mdx-js/.github/blob/HEAD/code-of-conduct.md)
- [X] I searched [issues](https://github.com/search?q=user%3Amdx-js&type=Issues) and couldn’t find anything (or linked relevant results below)

### Affected packages and versions

Latest

### Link to runnable example

https://stackblitz.com/edit/node-cppxmm?file=index.mdx

### Steps to reproduce

1. Open reproducible link
2. run `npx eslint ./index.mdx`

### Expected behavior

No error reported.

### Actual behavior

```
/home/projects/node-cppxmm/index.mdx
  1:8  error  'MyForm' is defined but never used  no-unused-vars
```

Enabling `eslint-plugin-react` would fix the error.

**Comment by chenxsan** [comment]
All right, here's why `eslint-plugin-react` would suppress the error:

1. https://github.com/jsx-eslint/eslint-plugin-react/blob/13d23b898de83557861b0e8bd419477f122e4839/lib/rules/jsx-uses-vars.js#L55

According to [react/jsx-uses-vars](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-uses-vars.md), ESLint has stopped detecting variables used in JSX.

Which means if we're going to use `eslint-plugin-mdx` without `eslint-plugin-react`, we will be sured to run into this `no-unused-vars` error.

Hence I would suggest adding a notice to the readme.

**Comment by kfayelun** [comment]
I have this issue, all my Storybook MDX components like `Meta`, `Canvas`, `Story` gives me a `no-unused-vars` error that I have downgraded to warning in my eslint config. I tried adding `eslint-plugin-react` to my config as described above, but that did not fix it. Is this the correct way to do that: 

```
module.exports = {
  root: true,
  extends: [
    "eslint:recommended",
    "plugin:storybook/recommended",
    "turbo",
    "prettier",
  ],
  env: {
    browser: true,
    node: true,
    es2021: true,
  },
  parserOptions: {
    ecmaVersion: "latest",
    sourceType: "module",
  },
  overrides: [
    {
      files: ["*.ts", "*.tsx"],
      parser: "@typescript-eslint/parser",
      extends: [
        "eslint:recommended",
        "plugin:@typescript-eslint/recommended",
        "plugin:storybook/recommended",
      ],
      plugins: ["@typescript-eslint"],
      rules: {
        "@typescript-eslint/no-var-requires": "warn",
      },
    },
    {
      files: ["*.md", "*.mdx"],
      extends: ["plugin:react/recommended", "plugin:mdx/recommended"],
      settings: {
        "mdx/code-blocks": false,
      },
      rules: {
        "no-unused-vars": "warn", // TODO: not sure how to fix this one for MDX
      },
    },
  ],
};
```

**Comment by JounQin** [comment]
@kfayelun Please provide a minimal and runnable reproduction.

**Comment by JounQin** [comment]
Maybe we should mark `eslint-plugin-react` as an optional peer dependency.

**Comment by kfayelun** [comment]
Hey! Working on a big monorepo, producing a runnable repro is quite the task so I had to put this on the backlog. And now... when I went back to check, this isn't happening anymore :( I've restructured so much in this repo, so I'm not sure what did it. I'm sorry, I know this is not very helpful.

**Comment by levino** [comment]
I have this or a very similar issue in one of my repos. [PR](https://github.com/levino/levinkeller.de/pull/17). [Failing CI Job](https://github.com/levino/levinkeller.de/actions/runs/5784598936/job/15675597072?pr=17). I guess the best way of action would be to get `@typescript-eslint` to understand the usage of imports in `.mdx` files. But I have not idea how complex that would be.

**Comment by JounQin** [comment]
@levino Sorry I didn't notice your comment previously, did you try to install `eslint-plugin-react` as this issue described?

**Comment by ropdias** [comment]
@lsarni Thanks for digging into this.

I reproduced the issue locally and found an important detail that explains the different behaviors we were seeing.

What was happening:
- With Flat Config, adding `plugins: { react }` after `...mdx.flat` replaces the entire `plugins` object.
- This removes the `mdx` plugin while keeping `mdx/*` rules in the config.
- In practice, this caused ESLint to either error out or effectively stop linting `.mdx` files, which made the original `no-unused-vars` issue appear “fixed” even though MDX wasn’t being linted correctly.

So the workaround fixed the symptom, but by accidentally disabling proper MDX linting.

What I changed to fix this properly:
1) **Merged `plugins` instead of replacing them**, so both `mdx` and `react` are registered in the same config object.
2) **Merged `rules` instead of replacing them**, preserving all `mdx.flat` rules.
3) **Added `react/jsx-uses-vars`**, so identifiers used only as JSX elements (e.g. `<Figure />`) are marked as “used” and don’t trigger `no-unused-vars`.

Working config (tested locally, ESLint runs on `.mdx` files correctly):

```js
import * as mdx from 'eslint-plugin-mdx'
import react from 'eslint-plugin-react'

export default [
  {
    files: ['**/*.mdx'],
    ...mdx.flat,
    plugins: {
      ...(mdx.flat.plugins ?? {}),
      react,
    },
    processor: mdx.createRemarkProcessor({ lintCodeBlocks: true }),
    rules: {
      ...(mdx.flat.rules ?? {}),
      'react/jsx-uses-vars': 'error',
    },
    settings: {
      react: { version: 'detect' },
    },
  },
]
```

Before calling this the final solution, could you please test this config on your setup and confirm if ESLint now runs correctly on .mdx files and the no-unused-vars false positive is gone?

If it does, I think this fully addresses the Flat Config + MDX case here.

**Comment by ropdias** [comment]
@lsarni Thanks for confirming 👍

Great to hear that with this config the “is defined but never used” error is gone. 🎉

And yeah, the Storybook import error (`@storybook/addon-docs/blocks`) is a separate issue and unrelated to the MDX + `no-unused-vars` problem. Adding the React plugin just makes ESLint run more completely, so it can surface unrelated resolution issues that were already present. I’d treat that as a separate configuration or Storybook-specific concern.

**Comment by controversial** [comment]
[eslint-react](https://github.com/Rel1cx/eslint-react/blob/main/CHANGELOG.md) is removing `jsx-uses-vars` in v3 because eslint 10 is supposed to [detect support for vars in JSX](https://eslint.org/docs/latest/use/migrate-to-10.0.0#-jsx-references-are-now-tracked). The ESLint 10 release says

> Rules previously used to work around ESLint’s lack of JSX reference tracking (for example, [@eslint-react/jsx-uses-vars](https://www.eslint-react.xyz/docs/rules/jsx-uses-vars)) are no longer needed. Remove or disable them in your configuration.

However, it seems like the MDX plugin still has false positives without `jsx-uses-vars`? Can this plugin be updated to take advantage of ESLint 10’s “native support” for tracking JSX references?

**Comment by controversial** [comment]
actually, it works nicely in eslint 10 with config like
```ts
    {
      files: ["**/*.mdx"],
      languageOptions: {
        parserOptions: {
          ecmaFeatures: { jsx: true },
        },
      },
    },
```
